apps/scratch/src/extensions/assertions/index.ts

```diff
@@ -0,0 +1,397 @@
+
+    this.enabled = true;
+    this.runtime.emit("ASSERTIONS_ENABLED");
+  };
+
+  /**
+   * Called when the project is loaded while assertions are enabled.
+   * This requires us to remember the initial state of all targets.
+   */
+  onProjectLoaded = (): void => {
+    this.rememberTargetStates();
+  };
+
+  /**
+   * Called when assertions are to be disabled.
+   * Stops resetting targets when the project starts running and
+   * no longer checks the assertions when the project stops running.
+   */
+  onDisableAssertions = (): void => {
+    this.runtime.off("targetWasCreated", this.onTargetCreated);
+    this.runtime.off("PROJECT_START", this.onProjectStart);
+    this.runtime.off("PROJECT_RUN_STOP", this.onProjectStop);
+    this.runtime.off("PROJECT_STOP_ALL", this.onProjectStopAll);
+
+    this.enabled = false;
+    this.runtime.emit("ASSERTIONS_DISABLED");
+  };
+
+  /**
+   * When a Target is cloned, clone the custom state.
+   * @param newTarget - the newly created target.
+   * @param sourceTarget - the target used as a source for the new clone, if any.
+   * @listens Runtime#event:targetWasCreated
+   * @private
+   */
+  onTargetCreated(
+    newTarget: VM.TargetExtended,
+    sourceTarget?: VM.TargetExtended,
+  ): void {
+    if (sourceTarget) {
+      const state = (
+        sourceTarget as VM.TargetWithCustomState<CustomState>
+      ).getCustomState(AssertionExtension.STATE_KEY);
+
+      if (state) {
+        (newTarget as VM.TargetWithCustomState<CustomState>).setCustomState(
+          AssertionExtension.STATE_KEY,
+          {
+            ...state,
+          },
+        );
+      }
+    }
+  }
+
+  /**
+   * Callback for when the project starts running.
+   */
+  onProjectStart = (): void => {
+    this.resetAllTargets();
+  };
+
+  /**
+   * Remembers the state of all targets so they can be reset later.
+   */
+  rememberTargetStates = (): void => {
+    this.rememberedSpriteState = {};
+    this.rememberedStageState = null;
+
+    for (const target of this.runtime.targets) {
+      if (target.isStage) {
+        this.rememberedStageState = {
+          currentCostume: target.currentCostume,
+          volume: target.volume,
+          tempo: target.tempo,
+          videoTransparency: target.videoTransparency,
+        };
+      } else if (target.isSprite()) {
+        this.rememberedSpriteState[target.id] = {
+          currentCostume: target.currentCostume,
+          volume: target.volume,
+          name: target.getName(),
+          x: target.x,
+          y: target.y,
+          size: target.size,
+          direction: target.direction,
+          draggable: target.draggable,
+          rotationStyle: target.rotationStyle,
+          layerOrder: target.getLayerOrder(),
+        };
+      }
+    }
+  };
+
+  /**
+   * Reset all targets to their initial state, i.e. the state they were in when assertions were enabled.
+   */
+  resetAllTargets = (): void => {
```

Comment:
This seems like a potentially the wrong approach. Why try to encompass all other possible state instead of working in our own tiny sandbox (e.g. `target.extensions_crt_assert.*`) and only reverting that state? 🤔 

I mean, sure, this prevents an assertion that does weird stuff from messing up the state, but I'd leave that as the responsibility of the task creator.

Reply:
as discussed in person: for custom state yes, for the other states we want to reset them after running the task because scratch does not do that otherwise.